client/src/main/java/io/pravega/client/tables/TableEntry.java

```diff
@@ -23,72 +24,110 @@
 
 /**
  * A {@link KeyValueTable} Entry.
- *
- * @param <KeyT>   Key Type.
- * @param <ValueT> Value Type
  */
 @Getter
 @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
 @ToString
-public class TableEntry<KeyT, ValueT> {
+public class TableEntry {
     /**
      * The {@link TableKey}.
      *
      * @return The content of {@link TableKey}.
      */
     @NonNull
-    private final TableKey<KeyT> key;
+    private final TableKey key;
 
     /**
      * The Value.
      *
      * @return The value associated to the {@link TableKey}.
      */
-    private final ValueT value;
+    private final ByteBuffer value;
+
+    /**
+     * Gets a value indicating whether this {@link TableEntry} has a specified {@link Version}.
+     *
+     * If versioned, then the following methods on {@link KeyValueTable} will be conditioned on this {@link TableEntry}'s
+     * {@link TableKey} matching the server-side version with that provided in this instance:
+     * - {@link KeyValueTable#put(TableEntry)}
+     * - {@link KeyValueTable#putAll(Iterable)}
+     *
+     * @return True if versioned, false otherwise.
+     */
+    public boolean isVersioned() {
+        return this.key.isVersioned();
+    }
+
+    /**
+     * Creates a new {@link TableEntry} with a Key made of only a Primary Key, with no specific version.
+     *
+     * @param primaryKey The Entry Key (Primary Key only).
+     * @param value      The Entry Value.
+     * @return An unversioned {@link TableEntry} (version set to {@link Version#NO_VERSION}).
+     */
+    public static TableEntry unversioned(ByteBuffer primaryKey, ByteBuffer value) {
+        return unversioned(primaryKey, null, value);
+    }
```

Comment:
Yes, I have considered that. However I can't think of an elegant way of dealing with it. The behavior of the KVT is dependent on the configuration that the user specified at runtime (when creating it). We can't force the user to choose different APIs based on whether they chose a secondary key or not, just like we can't fix key-length problems at compile time. 

This will be a runtime error and the request will be rejected with the appropriate error message in this case. I believe this is acceptable.

Reply:
This method has been removed.